cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,529 @@
+          + "a config error. Topics should not be excluded=[%s]", String.join(", ", shouldNotBeExcludedTopics)));
+    }
+  }
+
+  private void validateEnoughLeaderReplicaToDistribute(ClusterModel clusterModel,
+                                                       OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    if (_mustHaveLeaderReplicaPerBrokerTopics.isEmpty()) {
+      return;
+    }
+    Map<String, Set<Replica>> replicasByTopicNames = clusterModel.replicasOf(_mustHaveLeaderReplicaPerBrokerTopics);
+    Set<Broker> eligibleBrokersForLeadership = eligibleToHaveLeaderReplicaBrokers(clusterModel, optimizationOptions);
+    int totalMinimumLeaderReplicaCount = eligibleBrokersForLeadership.size() * minTopicLeadersPerBroker();
+
+    for (Map.Entry<String, Set<Replica>> replicasByTopicName : replicasByTopicNames.entrySet()) {
+      int totalLeaderReplicaCountForTopic = countLeaderReplicas(replicasByTopicName.getValue());
+      if (totalLeaderReplicaCountForTopic < totalMinimumLeaderReplicaCount) {
+        throw new OptimizationFailureException(
+            String.format("Cannot distribute %d leader replica(s) over %d broker(s) with total minimum required leader "
+                    + "replica count %d for topic %s.",
+                totalLeaderReplicaCountForTopic, eligibleBrokersForLeadership.size(),
+                totalLeaderReplicaCountForTopic, replicasByTopicName.getKey()));
+      }
+    }
+  }
+
+  private int countLeaderReplicas(Collection<Replica> replicas) {
+    int leaderReplicaCount = 0;
+    for (Replica replica : replicas) {
+      if (replica.isLeader()) {
+        leaderReplicaCount++;
+      }
+    }
+    return leaderReplicaCount;
+  }
+
+  private void validateBrokersAllowedReplicaMoveExist(ClusterModel clusterModel,
+                                                      OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    Set<Integer> brokersAllowedReplicaMove = GoalUtils.aliveBrokersNotExcludedForReplicaMove(clusterModel, optimizationOptions);
+    if (brokersAllowedReplicaMove.isEmpty()) {
+      // Handle the case when all alive brokers are excluded from replica moves.
+      throw new OptimizationFailureException("Cannot take any action as all alive brokers are excluded from replica moves.");
+    }
+  }
+
+  /**
+   * Check if requirements of this goal are not violated if this action is applied to the given cluster state,
+   * false otherwise.
+   *
+   * @param clusterModel The state of the cluster.
+   * @param action Action containing information about potential modification to the given cluster model.
+   * @return True if requirements of this goal are not violated if this action is applied to the given cluster state,
+   * false otherwise.
+   */
+  @Override
+  protected boolean selfSatisfied(ClusterModel clusterModel, BalancingAction action) {
+    if (!_mustHaveLeaderReplicaPerBrokerTopics.contains(action.topic())) {
```

Comment:
If this is a replica swap operation, then this check might miss destination topic that might be relevant to this goal. This may cause accepting a replica swap that it was supposed to reject.

Reply:
Thanks for pointing out this point. I assumed that the swap operation always swaps replicas of the same topic. So, I did not consider the case in which two to-be-swapped topics could be different.

I add some logic to handle this case.